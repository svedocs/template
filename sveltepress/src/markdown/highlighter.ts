import type { Highlighter } from '@sveltepress/vite'
import type { BundledTheme, CodeToHastOptions, HighlighterGeneric, ShikiTransformer } from 'shiki'
import type { BundledLanguage } from 'shiki/langs'
import type { DefaultThemeOptions } from 'virtual:sveltepress/theme-default'
import { env } from 'node:process'
import { createTransformerFactory } from '@shikijs/twoslash'
import { createTwoslasher, rendererFloatingSvelte } from '@sveltepress/twoslash'
import { LRUCache } from 'lru-cache'
import { createHighlighter } from 'shiki'
import { themeOptionsRef } from '../index.js'
import { processCommands } from './commands.js'

const DEFAULT_SUPPORT_LANGUAGES: any[] = ['svelte', 'sh', 'js', 'html', 'ts', 'md', 'css', 'scss']

const cache = new LRUCache<string, any>({ max: 200 })

const shikiHighlighterInstance: {
  value: null | HighlighterGeneric<BundledLanguage, BundledTheme>
  twoslashTransformer?: ShikiTransformer
} = {
  value: null,
}

const _highlighter: Highlighter = async (code, lang) => {
  const highlighterConfig = themeOptionsRef.value?.highlighter
  const darkTheme = highlighterConfig?.themeDark ?? 'night-owl'
  const lightTheme = highlighterConfig?.themeLight ?? 'vitesse-light'

  const shikiOptions: CodeToHastOptions<BundledLanguage, BundledTheme> = {
    lang,
    themes: {
      dark: darkTheme,
      light: lightTheme,
    },
    transformers: [],
  }
  if (shikiHighlighterInstance.twoslashTransformer) {
    shikiOptions.transformers?.push(
      shikiHighlighterInstance.twoslashTransformer,
    )
  }
  const { codeToHtml } = shikiHighlighterInstance.value!
  return codeToHtml(code, shikiOptions)
    .replace(/\{/g, '&#123;')
    .replace(/\}/g, '&#125;')
    .replace(/<!--svp-floating-snippet-start-->/g, '{#snippet floatingContent()}')
    .replace(/<!--svp-floating-snippet-end-->/g, '{/snippet}')
}

const highlighter: Highlighter = async (code, lang, meta) => {
  const cacheKey = JSON.stringify({ code, lang, meta })
  let cached
  if (env.NODE_ENV === 'development') {
    cached = cache.get(cacheKey)
    if (cached)
      return cached
  }
  const metaArray = (meta || '').split(' ')
  const containLineNumbers = metaArray.some(item => item.trim() === 'ln')
  const titleMeta = metaArray.find(item => item.startsWith('title='))
  const commandDoms: string[] = []
  const lines = code.split('\n')
  let noErrorsFirstLine: string | undefined
  if (lines[0] === '// @noErrors')
    noErrorsFirstLine = lines.shift()
  code = lines.map((line, i) => {
    const [commandDomsInOneLine, newLine] = processCommands(line, i, lines.length)
    commandDoms.push(...commandDomsInOneLine)
    return newLine
  }).join('\n')
  let title: string | undefined
  if (titleMeta)
    title = titleMeta.split('=')[1].replace(/(^")|("$)/g, '')
  if (noErrorsFirstLine)
    code = `${noErrorsFirstLine}\n${code}`

  cached = `
<div class="svp-code-block-wrapper">${title
  ? `<div class="svp-code-block--title">${title}</div>
`
  : ''}
  <div class="svp-code-block${containLineNumbers ? ' svp-code-block--with-line-numbers' : ''}">
    ${commandDoms.join('\n')}
    <!-- svelte-ignore a11y_no_noninteractive_tabindex -->
    ${await _highlighter(code, lang, meta)}
    <div class="svp-code-block--lang">
      ${lang}
    </div>
    <CopyCode />
    ${containLineNumbers ? `<div class="svp-code-block--line-numbers">${lines.map((_, i) => `<div class="svp-code-block--line-number-item">${i + 1}</div>`).join('\n')}</div>` : ''}
  </div>
</div>`
  if (env.NODE_ENV === 'development')
    cache.set(cacheKey, cached)
  return cached
}

export async function initHighlighter(highlighterConfig: DefaultThemeOptions['highlighter'] = {}) {
  const langs = highlighterConfig?.languages || DEFAULT_SUPPORT_LANGUAGES
  const darkTheme = highlighterConfig?.themeDark ?? 'night-owl'
  const lightTheme = highlighterConfig?.themeLight ?? 'vitesse-light'
  shikiHighlighterInstance.value = await createHighlighter({
    langs,
    themes: [darkTheme, lightTheme],
  })
  if (highlighterConfig.twoslash) {
    shikiHighlighterInstance.twoslashTransformer = createTransformerFactory(
      await createTwoslasher(highlighterConfig.twoslash === true
        ? {
            compilerOptions: {
              module: 199,
              moduleResolution: 99,
              jsx: 1,
              types: ['@sveltepress/vite/types', '@sveltepress/theme-default/types', '@sveltepress/theme-default/components', '@sveltejs/kit'],
            },
          }
        : highlighterConfig.twoslash),
    )(
      {
        langs: ['ts', 'tsx', 'svelte'],
        renderer: rendererFloatingSvelte(),
      },
    )
  }
}

export default highlighter
