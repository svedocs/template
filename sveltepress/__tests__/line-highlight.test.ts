import { describe, expect, it } from 'vitest'
import highlighter, { initHighlighter } from '../src/markdown/highlighter'

describe('line highlight', async () => {
  await initHighlighter()

  it('single line', async () => {
    const source = `<script>
  let count = $state(0) // [svp! hl]
</script>

<button onclick="{() => count++}">
  You've clicked {count} times.
</button>
`
    expect(await highlighter(source, 'svelte')).toMatchSnapshot()
  })
})
