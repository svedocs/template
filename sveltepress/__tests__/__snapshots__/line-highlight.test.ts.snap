// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`line highlight > single line 1`] = `
"
<div class="svp-code-block-wrapper">
  <div class="svp-code-block">
    <div class="svp-code-block--command-line svp-code-block--hl"  style="top: calc(1.5em + 12px);"></div>
    <!-- svelte-ignore a11y_no_noninteractive_tabindex -->
    <pre class="shiki shiki-themes vitesse-light night-owl" style="background-color:#ffffff;--shiki-dark-bg:#011627;color:#393a34;--shiki-dark:#d6deeb" tabindex="0"><code><span class="line"><span style="color:#999999;--shiki-dark:#7FDBCA">&#x3C;</span><span style="color:#1E754F;--shiki-dark:#CAECE6">script</span><span style="color:#999999;--shiki-dark:#7FDBCA">></span></span>
<span class="line"><span style="color:#AB5959;--shiki-dark:#C792EA">  let</span><span style="color:#B07D48;--shiki-dark:#D7DBE0"> count</span><span style="color:#999999;--shiki-dark:#C792EA"> =</span><span style="color:#999999;--shiki-dark:#C792EA"> $</span><span style="color:#59873A;font-style:inherit;--shiki-dark:#82AAFF;--shiki-dark-font-style:italic">state</span><span style="color:#999999;--shiki-dark:#D6DEEB">(</span><span style="color:#2F798A;--shiki-dark:#F78C6C">0</span><span style="color:#999999;--shiki-dark:#D6DEEB">)</span><span style="color:#393A34;font-style:inherit;--shiki-dark:#C792EA;--shiki-dark-font-style:italic"> </span></span>
<span class="line"><span style="color:#999999;--shiki-dark:#7FDBCA">&#x3C;/</span><span style="color:#1E754F;--shiki-dark:#CAECE6">script</span><span style="color:#999999;--shiki-dark:#7FDBCA">></span></span>
<span class="line"></span>
<span class="line"><span style="color:#999999;--shiki-dark:#7FDBCA">&#x3C;</span><span style="color:#1E754F;--shiki-dark:#CAECE6">button</span><span style="color:#B07D48;font-style:inherit;--shiki-dark:#C5E478;--shiki-dark-font-style:italic"> onclick</span><span style="color:#999999;--shiki-dark:#7FDBCA">=</span><span style="color:#B5695977;--shiki-dark:#D9F5DD">"</span><span style="color:#999999;--shiki-dark:#D3423E">&#123;</span><span style="color:#999999;--shiki-dark:#D9F5DD">()</span><span style="color:#999999;--shiki-dark:#C792EA"> =></span><span style="color:#B56959;--shiki-dark:#D6DEEB"> count</span><span style="color:#AB5959;--shiki-dark:#C792EA">++</span><span style="color:#999999;--shiki-dark:#D3423E">&#125;</span><span style="color:#B5695977;--shiki-dark:#D9F5DD">"</span><span style="color:#999999;--shiki-dark:#7FDBCA">></span></span>
<span class="line"><span style="color:#393A34;--shiki-dark:#D6DEEB">  You've clicked </span><span style="color:#999999;--shiki-dark:#D3423E">&#123;</span><span style="color:#B07D48;--shiki-dark:#D6DEEB">count</span><span style="color:#999999;--shiki-dark:#D3423E">&#125;</span><span style="color:#393A34;--shiki-dark:#D6DEEB"> times.</span></span>
<span class="line"><span style="color:#999999;--shiki-dark:#7FDBCA">&#x3C;/</span><span style="color:#1E754F;--shiki-dark:#CAECE6">button</span><span style="color:#999999;--shiki-dark:#7FDBCA">></span></span>
<span class="line"></span></code></pre>
    <div class="svp-code-block--lang">
      svelte
    </div>
    <CopyCode />
    
  </div>
</div>"
`;
